Cython/Compiler/PyrexTypes.py

```diff
@@ -4773,6 +4783,14 @@ def specialize_here(self, pos, env, template_values=None):
         # Replace this type with the actual 'template' argument.
         return template_values[0].resolve()
 
+class PyNoneType(BuiltinObjectType):
```

Comment:
Unless there's a good reason for this that I'm missing, I'd declare this with all the other builtin types in Cython/Compiler/Builtins.py.

It isn't obvious to me where the `__bool__` overload is actually used?

Reply:
> Unless there's a good reason for this that I'm missing, I'd declare this with all the other builtin types in Cython/Compiler/Builtins.py.

No reason except that I am still newbie in Cython codebase. I will move it.

> It isn't obvious to me where the `__bool__` overload is actually used?

In cython code there are parts where  you have line `if typevar:` (at least in code analysing binary operators or assignment) where in master `typevar` contains `None` but in this PR it is `PyNoneType`. `__bool__` ensures the `if` statements will be evaluated in the same way than before. If you want concrete example just comment out `__bool__` method and run tests - you will get test failures.

Reply:
> In cython code there are parts where you have line `if typevar:`

E.g. this line code:
https://github.com/cython/cython/blob/7e2cd6ad57c9eeb1fc6bf5d2bea469888a3f9b61/Cython/Compiler/ExprNodes.py#L12142

Reply:
> `if type1 and ...`

Typically that'd be testing that `type1 is not None` which would mean it hasn't yet been analysed. That's different from "`type1` has been analysed and is known to be none_type".

I'll have a proper look later this week and try to understand where that's making a difference.